runtime/caml/eventring.h

```diff
@@ -0,0 +1,296 @@
+             ev_runtime_phase phase));
+
+/* Set the runtime_counter event callback on the cursor */
+extern void caml_eventring_set_runtime_counter(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, uint64_t timestamp,
+             ev_runtime_counter counter, uint64_t val));
+
+/* Set the alloc event callback on the cursor */
+extern void
+caml_eventring_set_runtime_alloc(struct caml_eventring_cursor *cursor,
+                                 int (*f)(int domain_id, void *callback_data,
+                                          uint64_t timestamp, uint64_t *sz));
+
+/* Set the lifecycle event callback on the cursor */
+extern void caml_eventring_set_lifecycle(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, int64_t timestamp,
+              ev_lifecycle lifecycle, int64_t data));
+
+/* Set the lost events callback on the cursor */
+extern void caml_eventring_set_lost_events(
+    struct caml_eventring_cursor *cursor,
+    int (*f)(int domain_id, void *callback_data, int lost_words));
+
+/* frees a cursor obtained from caml_eventring_creator_cursor */
+extern void
+caml_eventring_free_cursor(struct caml_eventring_cursor *cursor);
+
+/* polls the eventring pointed to by [cursor] and calls the appropriate callback
+    for each new event up to at most [max_events] times.
+
+    Returns the number of events consumed in [events_consumed], if set.
+
+    0 for [max_events] indicates no limit to the number of callbacks. */
+CAMLextern eventring_error caml_eventring_read_poll(
+    struct caml_eventring_cursor *cursor,
+    void *callback_data,
+    uintnat max_events, uintnat *events_consumed);
+
+/* OCaml API for reading from the eventring. Documented in eventring.ml */
+extern value caml_eventring_create_cursor_ml(value path_pid);
+extern value caml_eventring_free_cursor_ml(value wrapped_cursor);
+extern value caml_eventring_read_poll_ml(value wrapped_cursor,
+                                              value callbacks,
+                                              value max_events_option);
+
+#ifdef CAML_INTERNALS
+
+struct eventring_buffer_header {
+  atomic_uint_fast64_t ring_head;
+  atomic_uint_fast64_t ring_tail;
+  uint64_t padding[6]; /* Padding so headers don't share cache lines */
+};
+
+struct eventring_metadata_header {
+  uint64_t version;
+  uint64_t max_domains;
+  uint64_t ring_header_size_bytes; /* Ring buffer header size (bytes) */
+  uint64_t ring_size_bytes; /* Ring data size (bytes) */
+  uint64_t ring_size_elements; /* Ring size in 64-bit elements */
+  uint64_t headers_offset; /* Offset from this struct to first header (bytes) */
+  uint64_t data_offset; /* Offset from this struct to first data (byte) */
+  uint64_t padding; /* Make the header a multiple of 64 bytes */
+};
+
+#define EVENTRING_NUM_ALLOC_BUCKETS 20
+#define EVENTRING_MAX_MSG_LENGTH (1 << 10)
+
+/* event header fields (for runtime events):
```

Comment:
This can describe as well user events according to the next line, so I guess the `(for runtime events)` part is not quite accurate?